crates/sol-types/src/types/calls.rs

```diff
@@ -0,0 +1,131 @@
+use crate::{Panic, Result, Revert, SolError};
+use alloc::vec::Vec;
+
+/// A collection of ABI-encoded call-like types. This currently includes
+/// [`SolCall`] and [`SolError`].
+///
+/// [`SolCall`]: crate::SolCall
+/// [`SolError`]: crate::SolError
+///
+/// ### Implementer's Guide
+///
+/// We do not recommend implementing this trait directly. Instead, we recommend
+/// using the [`sol`][crate::sol] proc macro to parse a Solidity contract
+/// definition.
+pub trait SolCalls: Sized {
```

Comment:
should we enable iterating over variant names/signatures/selectors?

Reply:
I initially had a `const SELECTORS: &'static [[u8; 4]]` but that doesn't work with the nested `ContractError` I later added

Reply:
What do you think of
```rust
trait {
  fn selectors() -> Selectors<Self> {}
  fn selector_at(i: usize) -> Option<[u8; 4]> {}
}

struct Selectors<T> { index: usize }
impl<T> Iterator for Selectors<T> { ... }
```

Reply:
I've addded the above interface, as well as re-adding the `const SELECTORS` as a associated constant on the enum itself, instead of on the trait

Reply:
I like it.